lib/pg/instrumentation.rb

```diff
@@ -0,0 +1,184 @@
+            @shared_tags['db.user'] = db_user if db_user
+            @shared_tags['peer.hostname'] = host if host
+            @shared_tags['peer.port'] = port if port
+            @shared_tags['peer.address'] = "pg://#{host}:#{port}" if host && port
+
+            operation_name = 'pg.initialize'
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: @shared_tags)
+
+            initialize_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def async_exec(*args)
+            tags = @shared_tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: tags)
+
+            async_exec_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def exec(*args)
+            tags = @shared_tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: tags)
+
+            exec_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def exec_params(*args)
+            tags = @shared_tags.dup
+
+            sql = args.first.to_s[0, 1024]
+            tags['db.statement'] = sql
+
+            default_op_name = 'pg.query'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: tags)
+
+            exec_params_original(*args)
+
+          rescue => e
+            log_error(scope.span, e) if scope
+          ensure
+            scope&.close
+          end
+
+          def prepare(*args)
+            tags = @shared_tags.dup
+            @prep_sql = args
+
+            sql = args[1].to_str[0, 1024]
+            tags['db.statement'] = sql
+            tags['prepared.statement.name'] = args.first
+
+            default_op_name = 'pg.prepare'
+            operation_name = get_operation_name(sql, default_op_name)
+
+            scope = ::PG::Instrumentation.tracer.start_active_span(operation_name, tags: tags)
+
+            prepare_original(*args)
+
+            rescue => e
+              log_error(scope.span, e) if scope
+            ensure
+              scope&.close
+            end
+
+          def exec_prepared(*args)
+            tags = @shared_tags.dup
+
+            if args.first == @prep_sql.first
```

Comment:
What is the significance of this?  In general though, are you sure it's necessary or wise to share state between calls instead of denoting the available information for each traced function? `async_` and connection pooling scenarios seem a bit concerning.

Reply:
I thought it was a "nice to have", not "must have". The aim was to get additional/clearer information on the exact sql statement being executed (variables included). However, I also debated just leaving it to the information available from the function being wrapped. Can take it out.